### PR TITLE
feat(storage): compare deserialized row value in sanity check

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Previously in local state store, for case `is_consistent_op=true`, we compare the serialized bytes of old value, which may fail the sanity check after schema check. In this PR, we change to pass a dynamic old value checker to perform the old value sanity check. In this way, we can create a dynamic old value checker in state table to deserialize the raw value bytes and compare the row value. 

We change the bool `is_consistent_op` in `NewLocalOptions` to an enum `OpConsistentLevel`. 

The state table of materialized executor has `is_consistent_op=true` even after schema check.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/store.rs [hunk 1]
```diff
@@ -394,6 +394,18 @@ impl From<TracedWriteOptions> for WriteOptions {
...
+
+pub static CHECK_BYTES_EQUAL: LazyLock<Arc<dyn CheckOldValueEquality>> =
+    LazyLock::new(|| Arc::new(|first: &Bytes, second: &Bytes| first == second));
+
+#[derive(Default, Clone)]
+pub enum OpConsistencyLevel {
+    #[default]
+    Inconsistent,
+    ConsistentOldValue(Arc<dyn CheckOldValueEquality>),
+}
```
Comment (BugenZhao) [inline src/storage/src/store.rs:407]: It appears that we only have two kinds of `impl Check` now: equal by bytes, or by values. Is it possible to make it an `enum` to get rid of `dyn` here?

File: src/stream/src/common/table/state_table.rs [hunk 2]
```diff
@@ -211,6 +211,31 @@ where
     }
 }
 
+fn consistent_old_value_op(row_serde: impl ValueRowSerde) -> OpConsistencyLevel {
+    OpConsistencyLevel::ConsistentOldValue(Arc::new(move |first: &Bytes, second: &Bytes| {
```
Comment (BugenZhao) [inline src/stream/src/common/table/state_table.rs:215]: What about first comparing the bytes, then the deserialized values only if we find differences?

**Review by BugenZhao — APPROVED** [review]
LGTM! IIUC, the sanity check is currently only enabled on `debug_assertions`. So the performance may not be a major concerning.

File: src/storage/src/store.rs (hunk 1 above)
Comment (wenym1) [reply src/storage/src/store.rs:407]: Since the performance is not a major concern, I prefer keeping it simple.

**Review by Li0k — COMMENTED** [review]
LGTM

**Pull request merged**